segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName) {
+            PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getFilesystemRoot(), streamSegmentName),
+                    PosixFileAttributes.class);
+            StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        boolean exists = Files.exists(Paths.get(config.getFilesystemRoot(), streamSegmentName));
+        return exists;
+    }
+
+    @SneakyThrows
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        log.info("Creating Segment {}", streamSegmentName);
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, null).get();
+    }
+
+    @SneakyThrows
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.warn("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            throw new IllegalArgumentException("Write called on a readonly handle of segment "
+                    + handle.getSegmentName());
+        }
+
+        //Fix for Jenkins as Jenkins runs as super user privileges.
+        if ( !isWritableFile(path)) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        long fileSize = path.toFile().length();
+        if (fileSize < offset) {
+            throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
+        } else {
+            try (FileChannel channel = FileChannel.open(path,
+                    StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                while (length != 0) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
```

Comment:
You could get stuck in an infinite loop here. According to the documentation of `transferFrom`, this could start returning 0. You need to handle that case.
https://docs.oracle.com/javase/8/docs/api/java/nio/channels/FileChannel.html#transferFrom-java.nio.channels.ReadableByteChannel-long-long-